### DISCUSSION TITLE
Can I insert if a value exists, otherwise use the default?

### DISCUSSION THREAD

**Discussion by Jengamon** [geldata/gel#4548] [Q&A]
I have an optional parameter, and I want to insert using it, but only if it exists.

**Comment by MiroslavPetrik** [comment]
Hey @Jengamon , you can test the optional param behavior easily with the REPL.

We will work with simple type
```
type Alert {
  property message -> str;
}
```

To execute some query with optional param we need:
- `$name` for param.
- each param needs explicit cast, here we have string so our cast is `<str>` and to make it optional we prefix it `<optional str>`
- last in case of empty param we use the [coalescing operator `??`](https://www.edgedb.com/docs/stdlib/set#operator::coalesce)
```
insert Alert { message := <optional str>$message ?? "default message" };
```

Next you will be asked for input, or leave it empty. So we press `Ctrl+D`:
```
Parameter <str>$message (Ctrl+D for empty set `{}`):
{default::Alert{id: b253b71c-673e-11ed-9a43-5ff9307e70f5}}
```

We got back the `uuid` of new Alert entry so we can verify, that the default was used:
```
select Alert {message} filter .id = <uuid>"b253b71c-673e-11ed-9a43-5ff9307e70f5";
{default::Alert{message: 'default message'}}
```

Note that schema default can't be used without the `?? "default value"`

e.g.
```
type Alert {
  required property message -> str {
    default := "schema default";
  };
}
```

By repeating the _insert without coalescing_ from above with the `Ctrl+D` we get the following error:

```
insert Alert { message := <optional str>$message };
Parameter <str>$message (Ctrl+D for empty set `{}`):
# we press Ctrl+D
edgedb error: MissingRequiredError: missing value for required property 'message' of object type 'default::Alert'
```

The schema default will be used when you don't use the `message` field in the insert query:

```
edgedb> insert Alert;
{default::Alert {id: c526c8fa-6740-11ed-852b-57578b175e42}}
edgedb> select Alert {message} filter .id = <uuid>"c526c8fa-6740-11ed-852b-57578b175e42";
{default::Alert {message: 'schema default'}}
```
